src/main/java/org/spongepowered/api/event/world/chunk/ChunkEvent.java

```diff
@@ -64,58 +66,131 @@ interface WorldScoped extends ChunkEvent {
     }
 
     /**
-     * Called when a {@link WorldChunk chunk} was unloaded.
+     * Called when a {@link WorldChunk chunk} is performing a block related operation.
      */
-    interface Unload extends WorldScoped {
+    interface Blocks extends WorldScoped {
 
         /**
-         * Called before the {@link WorldChunk chunk} is unloaded.
+         * Called when a block data of {@link WorldChunk chunk} is loaded.
+         * This can be called outside the {@link World#engine() main} {@link Thread}.
+         * It is NOT safe to perform modifications to the {@link World} or via
+         * {@link org.spongepowered.api.world.server.ServerLocation} as this could
+         * result in a deadlock.
          */
-        interface Pre extends Unload {
+        interface Load extends Blocks {
 
             /**
-             * Gets the {@link WorldChunk chunk} being changed.
+             * Gets the {@link WorldChunk chunk} block volume.
              *
-             * @return The chunk
+             * @return The block volume
              */
-            WorldChunk chunk();
+            BlockChunk blockVolume();
         }
 
         /**
-         * Called after the {@link WorldChunk chunk} is unloaded.
+         * Called when a {@link WorldChunk chunk} is performing a block related save.
          */
-        interface Post extends Unload {
+        interface Save extends Blocks {
+
+            /**
+             * Called before the {@link WorldChunk chunk} block data is saved. Cancelling this
+             * will prevent any of the chunk's block data being written to it's storage container.
+             */
+            interface Pre extends Save, Cancellable {
 
+                /**
+                 * Gets the {@link WorldChunk chunk} block volume.
+                 *
+                 * @return The block volume
+                 */
+                BlockChunk blockVolume();
+            }
+
+            /**
+             * Called after the {@link WorldChunk chunk} block data is saved.
+             * Guaranteed to exist in the chunk's block storage container.
+             * <p>
+             * Depending on the implementation, this event may be called off-thread.
+             */
+            interface Post extends Save {}
         }
     }
 
     /**
-     * Called when a {@link WorldChunk chunk} is performing a save.
+     * Called when a {@link WorldChunk chunk} is performing a entity related operation.
      */
-    interface Save extends WorldScoped {
+    interface Entities extends WorldScoped {
+
+        /**
+         * Called when an entity data of {@link WorldChunk chunk} is loaded.
+         * This can be called outside the {@link World#engine() main} {@link Thread}.
+         * It is NOT safe to perform modifications to the {@link World} or via
+         * {@link org.spongepowered.api.world.server.ServerLocation} as this could
+         * result in a deadlock.
+         */
+        interface Load extends Entities {
```

Comment:
What is the rammification of calling `spawnEntity` on the `EntityChunk`?

Reply:
That works and is safe. The entity is appended to be part of the list of entities to be loaded.